src/main/scala/workflow/graph/GraphExecution.scala

```diff
@@ -0,0 +1,66 @@
+package workflow.graph
+
+// rough idea given incrementalism: do everything it can in the base executor (which may be shared w/ other things) w/o inserting sources.
+// then create a "final executor" that is the base one w/ sources inserted, and optimized using the EquivalentNodeMerge optimizer.
+// The final value execution happens on that "final executor"
+// This two stage process allows "intuitive things" to happen a source being passed in is already processed elsewhere in the pipeline (e.g. making sure to reuse a cacher),
+// while pipeline fitting results can be reused across multiple pipeline applies, as they all share the same base executor.
+abstract class GraphExecution[T](
+    executor: GraphExecutor,
+    sources: Map[SourceId, Operator],
+    sink: SinkId,
+    expressionToOutput: Expression => T
+  ) {
+  private var _executor: GraphExecutor = executor
+  private var _sources: Map[SourceId, Operator] = sources
+  private var _sink: SinkId = sink
+
+  private[graph] def setExecutor(executor: GraphExecutor): Unit = {
+    this._executor = executor
+  }
+
+  private[graph] def setSources(sources: Map[SourceId, Operator]): Unit = {
+    this._sources = sources
+  }
+
+  private[graph] def setSink(sink: SinkId): Unit = {
+    this._sink = sink
+  }
+
+  private[graph] def getSources: Map[SourceId, Operator] = _sources
+  private[graph] def getSink: SinkId = _sink
+
+  private var ranExecution: Boolean = false
+  private lazy val finalExecutor: GraphExecutor = {
+    if (getSources.nonEmpty) {
+      _executor.partialExecute(getSink)
+
+      val (newGraph, newState) = EquivalentNodeMergeOptimizer.execute(getGraph, getState)
```

Comment:
So here is an explanation at a high level while I write the docs for this class.
In order for pipeline fitting to be carried through across `pipeline.apply` calls, and to ensure common subexpressions get eliminated when we do something like pass training data into a pipeline (when there is a cacher in the featurization steps), we do this GraphExecution as a two step process:

The first fits/optimizes/partially executes the GraphExecutor shared across all `pipeline.apply` calls as far as possible BEFORE inserting any sources. Then, it inserts the sources into the graph, does common subexpression elimination, and produces a `finalExecutor` used ONLY by that specific `pipeline.apply(data)` execution.

This two-step process allows us to fit a pipeline and share the fitted results across all `pipeline.apply` calls, AND not do unnecessary work when a pipeline is applied to it's own training data/ something we would reasonably expect to have been cached.